eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/Imp/UtxowSpec/Invalid.hs

```diff
@@ -2,228 +2,347 @@
-
-  it "Missing 2-phase script witness" $ do
-    const $ pendingWith "not implemented yet"
-
-  it "Redeemer with incorrect label" $ do
-    const $ pendingWith "not implemented yet"
+    submitFailingTx tx [injectFailure $ ScriptWitnessNotValidatingUTXOW [scriptHash]]
 
-  it "Phase 1 script failure" $ do
-    const $ pendingWith "not implemented yet"
+  let resetAddrWits tx = updateAddrTxWits $ tx & witsTxL . addrTxWitsL .~ []
+      fixupResetAddrWits = fixupPPHash >=> resetAddrWits
+      -- Would be an Iso if microlens supported Iso's
+      redeemersL :: Lens' (Redeemers era) (Map.Map (PlutusPurpose AsIx era) (Data era, ExUnits))
+      redeemersL = lens unRedeemers (const $ Redeemers @era)
+      -- PlutusPurpose serialization wasn't fixed until Conway
+      withPPRTFailures =
+        if eraProtVerLow @era < natVersion @9
+          then withCborRoundTripFailures
+          else id
+
+  forM_ (eraLanguages @era) $ \lang ->
+    withSLanguage lang $ \slang ->
+      describe (show lang) $ do
+        let redeemerSameAsDatumHash = hashPlutusScript $ redeemerSameAsDatum slang
+            alwaysSucceedsWithDatumHash = hashPlutusScript $ alwaysSucceedsWithDatum slang
+            alwaysSucceedsNoDatumHash = hashPlutusScript $ alwaysSucceedsNoDatum slang
+
+        it "MissingRedeemers" $ do
+          let scriptHash = redeemerSameAsDatumHash
+          txIn <- produceScript scriptHash
+          let missingRedeemer = mkSpendingPurpose $ AsItem txIn
+          let tx = mkBasicTx mkBasicTxBody & bodyTxL . inputsTxBodyL .~ [txIn]
+          withPostFixup (fixupResetAddrWits . (witsTxL . rdmrsTxWitsL .~ mempty)) $
+            submitFailingTx
+              tx
+              [ injectFailure $
+                  MissingRedeemers [(missingRedeemer, scriptHash)]
+              , injectFailure $
+                  CollectErrors [NoRedeemer missingRedeemer]
+              ]
+
+        it "MissingRequiredDatums" $ do
+          txIn <- produceScript redeemerSameAsDatumHash
+          let tx = mkBasicTx mkBasicTxBody & bodyTxL . inputsTxBodyL .~ [txIn]
+          let missingDatum = hashData @era (Data (P.I 3))
+          withPostFixup (fixupResetAddrWits . (witsTxL . datsTxWitsL .~ mempty)) $
+            submitFailingTx
+              tx
+              [injectFailure $ MissingRequiredDatums [missingDatum] []]
+
+        it "NotAllowedSupplementalDatums" $ do
+          txIn <- produceScript redeemerSameAsDatumHash
+          let extraDatumHash = hashData @era (Data (P.I 30))
+          let extraDatum = Data (P.I 30)
+          let tx =
+                mkBasicTx mkBasicTxBody
+                  & bodyTxL . inputsTxBodyL .~ [txIn]
+                  & witsTxL . datsTxWitsL .~ TxDats (Map.singleton extraDatumHash extraDatum)
+          submitFailingTx
+            tx
+            [injectFailure $ NotAllowedSupplementalDatums [extraDatumHash] []]
+
+        describe "PPViewHashesDontMatch" $ do
+          let
+            test hash = do
```

Comment:
Since the `let` binding is local to the `describe` block, I thought it was unnecessary to use a longer name. However, let's make things as easy as possible for future readers.